third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,929 @@
+// may change/disappear, but not until the scope is exited.  So the "possibly modified" property
+// shouldn't be trusted after that.
+fn find_possibly_modified_vars(
+    env: &GlobalEnv,
+    params: &[Parameter],
+    exp: &ExpData,
+) -> BTreeSet<(Symbol, Option<NodeId>)> {
+    let mut visiting_binding: ScopedMap<Symbol, NodeId> = ScopedMap::new();
+    let mut unsafe_variables: BTreeSet<(Symbol, Option<NodeId>)> = BTreeSet::new();
+
+    // Track when we are in a modifying scope.
+    let mut modifying = false;
+    // Use a stack to keep the modification status properly scoped.
+    let mut modifying_stack = Vec::new();
+    exp.visit_positions(&mut |pos, e| {
+        use ExpData::*;
+        match e {
+            Invalid(_) | Value(..) | LoopCont(..) => {
+                // Nothing happens inside these expressions, so don't modify `modifying` state.
+            },
+            LocalVar(id, sym) => {
+                let current_binding_id = visiting_binding.get(sym);
+                if modifying {
+                    trace!(
+                        "Var {} in binding {:?} used in node {} is unsafe",
+                        sym.display(env.symbol_pool()),
+                        current_binding_id,
+                        id.as_usize(),
+                    );
+                    unsafe_variables.insert((*sym, current_binding_id.copied()));
+                } else {
+                    trace!(
+                        "Var {} in binding {:?} used in node {} is ok",
+                        sym.display(env.symbol_pool()),
+                        current_binding_id,
+                        id.as_usize(),
+                    );
+                }
+            },
+            Temporary(id, idx) => {
+                if let Some(sym) = params.get(*idx).map(|p| p.0) {
+                    if modifying {
+                        let current_binding_id = visiting_binding.get(&sym);
+                        trace!(
+                            "Temp {} = Var {} in binding {:?} is unsafe",
+                            *idx,
+                            sym.display(env.symbol_pool()),
+                            current_binding_id
+                        );
+                        assert!(current_binding_id.is_none());
+                        unsafe_variables.insert((sym, None));
+                    };
+                } else {
+                    let loc = env.get_node_loc(*id);
+                    env.diag(
+                        Severity::Bug,
+                        &loc,
+                        &format!("Use of temporary with no corresponding parameter `{}`", idx),
+                    );
+                }
+            },
+            Call(_, op, _explist) => {
+                match op {
+                    // NOTE: we don't consider values in globals, so no need to
+                    // consider BorrowGlobal(ReferenceKind::Mutable)} here.
+                    Operation::Borrow(ReferenceKind::Mutable) => {
```

Comment:
I'm wondering why the complexity here. The only way how a variable can be mutated is if it is directly borrowed or if it assigned. There is no other way (in Move and in Rust too). So you could simplify this logic here significantly (similar as in the lambda lifter). What am I missing?

Reply:
As we discussed, the transfomation `&(x+0)` to `&x` is not valid, so is `&{x}` to `&x`. IMO its not adequate to mark `x` as unsafe, this is not what is really happening here. One way to deal with this might be to just always put the simplified expression in a block (the AST output may look ugly, but you can imagine an easy post-transformation which removes `{x}` if its not a parameter of a borrow operation.

Reply:
Please look at the current code.  This is now handled by the `in_mut_borrow` flag in `ExpRewriterFunctions for SimplifierRewriter<'env>::rewrite_exp` .  Notably, `protected_rexp` is created from `rexp` and wraps it in a protective `Sequence` if  we are looking at `x` which started out `{x}` and the context is `&` or `&{x}.f`

Reply:
What I don't understand is why you are still having those `modifying` set of vars. Variables are not modified unless they are directly borrowed. Why do we have all this complexity here instead of a simple traversal which collects the vars which are borrowed?